### PR TITLE
Documentation for the Democracy Module

### PR DESCRIPTION
**Pull request by ltfschoen**
Relates to https://github.com/paritytech/substrate-developer-hub/issues/28

Requires update for consistency with https://github.com/paritytech/substrate/pull/2172

### CODE REVIEW

**Comment by joepetrowski** [comment]
A lot of code exceeded our style guide length limit. Fixed in most recent commit, but can be rolled back if we want this to be strictly comments.

**Comment by ltfschoen** [comment]
@joepetrowski the indentation didn't look right in the code snippet when i ran ` cargo doc --package srml-democracy --open`. it was ok on lines where indentation had been done using spaces, but not when indentation had been done using tabs, so i just removed the tabs and replaced them with spaces

File: srml/democracy/src/vote_threshold.rs [hunk 1]
```diff
@@ -22,13 +22,13 @@ use parity_codec::{Encode, Decode};
...
 
-/// A means of determining if a vote is past pass threshold.
+/// A means of determining if a vote is past its pass threshold.
 #[derive(Clone, Copy, PartialEq, Eq, Encode, Decode)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
 pub enum VoteThreshold {
 	/// A supermajority of approvals is needed to pass this vote.
 	SuperMajorityApprove,
-	/// A supermajority of rejects is needed to fail this vote.
+	/// A supermajority of rejections is needed to fail this vote.
```
Comment (DemiMarie-temp) [inline srml/democracy/src/vote_threshold.rs]: How large a supermajority is needed?  This should be documented.

Comment (joepetrowski) [reply srml/democracy/src/vote_threshold.rs]: See update

**Comment by ltfschoen** [comment]
i've received feedback from rpmeier suggesting that we providing information about what origin the proposed call is executed with

**Comment by ltfschoen** [comment]
@shawntabrizi why are you force pushing and deleting all the commit history from other contributors? why not just close this PR and create a fresh branch of your own?

**Comment by shawntabrizi** [comment]
@ltfschoen I am not intending to discredit any work here, but these docs have been stale and were originally opened to merge to the master branch.

Since the SRML Democracy module has changed quite a bit in master, we re-targeted this branch to merge to v1.0.

However, at certain points this branch was re-based or merged with the master branch which means that when we changed the base branch to v1.0, we were getting changes on the order of 400+ files changed.

I isolated the changes and fixed up the PR. There could probably be a better way to do this, but end of the day, I just want to get these docs merged in.

**Comment by ltfschoen** [comment]
Fair enough,  thanks for cleaning it up!

File: srml/democracy/src/lib.rs [hunk 2]
```diff
@@ -14,7 +14,143 @@
...
+//!  are stored in `VotersFor` and `VoteOf`, respectively.
+//! - **Vote delegation:** Voters may delegate and undelegate their votes for an amount of lock periods.
+//! - **Proxy account:** Stash accounts (see the [Staking module](../srml_staking/index.html)) may add or
+//!  remove a proxy account to vote on a public referendum on behalf of the stash account.
+//! - **Vote validity:** The referendum being voted on must be an active referendum index of the
+//!  `ReferendumInfoOf` mapping. The voter (transactor) must have a balance above zero to signal approval.
+//! - **Maturity:** A mature referendum is one that expires at the current block.
+//! - **Vote tallying, passing, and execution:** When a block is finalized, the Democracy module will search
+//!  for mature referenda and tally their votes, then remove them from the Table of Referenda. If a referendum's
+//!  vote tally meets its vote threshold, then it will be passed and executed. Last, increment `NextTally`
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//!  vote tally meets its vote threshold, then it will be passed and executed. Lastly, `NextTally` is incremented
```

File: srml/democracy/src/lib.rs [hunk 3]
```diff
@@ -14,7 +14,143 @@
...
+//! - Tally votes of public referenda.
+//! - Pass and execute maturing public referenda.
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Functions
+//!
+//! - `propose` - Propose a sensitive action to be taken.
+//! - `second` - Second (sponsor) a proposal.
+//! - `vote` - Vote in a referendum.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `vote` - Vote on a referendum.
```

File: srml/democracy/src/lib.rs [hunk 4]
```diff
@@ -14,7 +14,143 @@
...
+//! - `proxy_vote` - Vote in a referendum on behalf of a stash.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `proxy_vote` - Vote on a referendum on behalf of a stash account.
```

File: srml/democracy/src/lib.rs [hunk 5]
```diff
@@ -14,7 +14,143 @@
...
+//! - `start_referendum` - Start a referendum.
+//! - `cancel_referendum` - Remove a referendum.
+//! - `cancel_queued` - Cancel a proposal queued for enactment.
+//! - `on_finalize` - Called when a block is finalized.
+//! - `set_proxy` - Specify a proxy. Called by the stash.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `set_proxy` - Specify a proxy account. Called by the stash account.
```

File: srml/democracy/src/lib.rs [hunk 6]
```diff
@@ -14,7 +14,143 @@
...
+//! - `resign_proxy` - Clear the proxy. Called by the proxy.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `resign_proxy` - Clear the proxy account. Called by the proxy account.
```

File: srml/democracy/src/lib.rs [hunk 7]
```diff
@@ -14,7 +14,143 @@
...
+//! - `remove_proxy` - Clear the proxy. Called by the stash.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `remove_proxy` - Clear the proxy account. Called by the stash account.
```

File: srml/democracy/src/lib.rs [hunk 8]
```diff
@@ -14,7 +14,143 @@
...
+//! - `delegate` - Delegate vote.
+//! - `undelegate` - Undelegate vote.
+//!
+//! ### Public Functions
+//!
+//! - `locked_for` - Get the balance locked in support of a proposal.
+//! - `is_active_referendum` - Return true if given index corresponds to an on-going referendum.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! - `is_active_referendum` - Return true if the given referendum index corresponds to an on-going referendum.
```

File: srml/democracy/src/lib.rs [hunk 9]
```diff
@@ -118,22 +254,26 @@ decl_module! {
...
-		/// Vote in a referendum. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// Vote in a referendum. If `vote.is_aye`, the vote is to enact the proposal;
 		/// otherwise it is a vote to keep the status quo.
 		fn vote(origin, #[compact] ref_index: ReferendumIndex, vote: Vote) -> Result {
 			let who = ensure_signed(origin)?;
 			Self::do_vote(who, ref_index, vote)
 		}
 
-		/// Vote in a referendum on behalf of a stash. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// Vote in a referendum on behalf of a stash. If `vote.is_aye`, the vote is to enact the proposal;
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
		/// Vote on a referendum using a proxy account on behalf of a stash account.. If `vote.is_aye`, the vote is to enact the proposal;
```

File: srml/democracy/src/lib.rs [hunk 10]
```diff
@@ -118,22 +254,26 @@ decl_module! {
 			<DepositOf<T>>::insert(proposal, deposit);
 		}
 
-		/// Vote in a referendum. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// Vote in a referendum. If `vote.is_aye`, the vote is to enact the proposal;
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
		/// Vote on a referendum. If `vote.is_aye`, the vote is to enact the proposal;
```

File: srml/democracy/src/lib.rs (hunk 5 above)
Comment (ltfschoen) [reply srml/democracy/src/lib.rs]: nitpick: if added consider updating implementation too https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R304

File: srml/democracy/src/lib.rs [hunk 11]
```diff
@@ -14,7 +14,143 @@
...
+//! use srml_support::{decl_module, dispatch::Result};
+//! # use srml_democracy as democracy;
+//! # use system::ensure_signed;
+//! # type VoteIndex = u64;
+//! pub trait Trait: democracy::Trait { }
+//!
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!
+//! 		fn proxy_set_approvals(origin, votes: Vec<bool>, index: VoteIndex) -> Result {
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]: nitpick: is this function indented too much?

File: srml/democracy/src/lib.rs [hunk 12]
```diff
@@ -14,7 +14,143 @@
...
+//! ```
+//! use srml_support::{decl_module, dispatch::Result};
+//! # use srml_democracy as democracy;
+//! # use system::ensure_signed;
+//! # type VoteIndex = u64;
+//! pub trait Trait: democracy::Trait { }
+//!
+//! decl_module! {
+//! 	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! 		/// Specify approval votes for a given referendum index. Called by a proxy account on behalf of a stash account.
```

File: srml/democracy/src/lib.rs (hunk 11 above)
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
//! 		fn set_approvals(origin, votes: Vec<bool>, index: ReferendumIndex) -> Result {
```

File: srml/democracy/src/lib.rs [hunk 13]
```diff
@@ -153,6 +293,8 @@ decl_module! {
 			<DispatchQueue<T>>::mutate(when, |items| if items.len() > which { items[which] = None });
 		}
 
+		/// Called when a block is finalized. Will launch a new referendum if it is time, finish
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
		/// Called when a block is finalized. Launch a new referendum if it is time to, finish
```

File: srml/democracy/src/lib.rs [hunk 14]
```diff
@@ -239,7 +398,8 @@ decl_storage! {
 		pub MinimumDeposit get(minimum_deposit) config(): BalanceOf<T>;
 		/// The delay before enactment for all public referenda.
 		pub PublicDelay get(public_delay) config(): T::BlockNumber;
-		/// The maximum number of additional lock periods a voter may offer to strengthen their vote. Multiples of `PublicDelay`.
+		/// The maximum number of additional lock periods a voter may offer to strengthen
+		/// their vote. Multiples of `PublicDelay`.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
		/// their vote. `PublicDelay` is a multiple of `LockPeriods`.
```

Comment (ltfschoen) [reply srml/democracy/src/lib.rs]: i don't think ""Multiples of `Public`" delay makes sense, since:
* `PublicDelay` is measured using `BlockNumber`, which is type `u64`.
* `LockPeriods` is type `i8`.

File: srml/democracy/src/lib.rs [hunk 15]
```diff
@@ -285,9 +445,9 @@ decl_event!(
 );
 
 impl<T: Trait> Module<T> {
-	// exposed immutables.
+	// Exposed immutables.
 
-	/// Get the amount locked in support of `proposal`; `None` if proposal isn't a valid proposal
+	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal.
```

Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
	/// Get the balance locked in support of `proposal`; `None` if proposal isn't a valid proposal index.
```

File: srml/democracy/src/lib.rs [hunk 16]
```diff
@@ -285,9 +445,9 @@ decl_event!(
...
 	/// index.
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
```

File: srml/democracy/src/lib.rs [hunk 17]
```diff
@@ -431,29 +615,44 @@ impl<T: Trait> Module<T> {
 		Self::deposit_event(RawEvent::Executed(index, ok));
 	}
 
+	/// Finds the proposal that will be the next referendum, unreserves the currency of
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
	/// Find the proposal that will be the next referendum, unreserves the currency of
```

Comment (ltfschoen) [reply srml/democracy/src/lib.rs]: nitpick: just for consistency, since we're saying `Tally` not `Tallying` in other doc comments

File: srml/democracy/src/lib.rs [hunk 18]
```diff
@@ -700,7 +906,7 @@ mod tests {
 			// 2 cannot fire 1's proxy:
 			assert_noop!(Democracy::remove_proxy(Origin::signed(2), 10), "wrong proxy");
 
-			// 1 fires his proxy:
+			// 1 fires their proxy:
```
Comment (ltfschoen) [inline srml/democracy/src/lib.rs]:
```suggestion
			// stash account 1 fires their proxy:
```

Comment (ltfschoen) [reply srml/democracy/src/lib.rs]: nitpick: i also think `their` should be replaced with `its`

File: srml/democracy/src/vote_threshold.rs [hunk 19]
```diff
@@ -22,13 +22,15 @@ use parity_codec::{Encode, Decode};
 use primitives::traits::{Zero, IntegerSquareRoot};
 use rstd::ops::{Add, Mul, Div, Rem};
 
-/// A means of determining if a vote is past pass threshold.
+/// A means of determining if a vote is past its pass threshold.
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
/// A means of determining if a vote is above the required threshold to pass.
```

File: srml/democracy/src/vote_threshold.rs [hunk 20]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
...
+		+ Copy
+{
+	/// Given:
+	///
+	/// - `approve` votes for a proposal,
+	/// - `against` votes against a proposal,
+	/// - a total electorate size of `electorate`, and
+	/// - the number of `voters` who voted,
+	///
+	/// this function will return true if the overall outcome is in favor of approval.
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
	/// Return true if the overall outcome is in favor of approval.
```

File: srml/democracy/src/vote_threshold.rs [hunk 21]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
...
+		+ Mul<Balance, Output = Balance>
+		+ Div<Balance, Output = Balance>
+		+ Rem<Balance, Output = Balance>
+		+ Copy
+{
+	/// Given:
+	///
+	/// - `approve` votes for a proposal,
+	/// - `against` votes against a proposal,
+	/// - a total electorate size of `electorate`, and
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
	/// - A total electorate size of `electorate`; and
```

File: srml/democracy/src/vote_threshold.rs [hunk 22]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
...
+		+ Ord
+		+ Add<Balance, Output = Balance>
+		+ Mul<Balance, Output = Balance>
+		+ Div<Balance, Output = Balance>
+		+ Rem<Balance, Output = Balance>
+		+ Copy
+{
+	/// Given:
+	///
+	/// - `approve` votes for a proposal,
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
	/// - Votes that `approve` a proposal.
```

File: srml/democracy/src/vote_threshold.rs [hunk 23]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
...
+	/// - `against` votes against a proposal,
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
	/// - Vote that are `against` a proposal.
```

File: srml/democracy/src/vote_threshold.rs [hunk 24]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
...
+	/// - a total electorate size of `electorate`, and
+	/// - the number of `voters` who voted,
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]:
```suggestion
	/// - Number of `voters` who voted.
```

File: srml/democracy/src/lib.rs (hunk 16 above)
Comment (ltfschoen) [reply srml/democracy/src/lib.rs]: i think this can fit on one line

File: srml/democracy/src/vote_threshold.rs [hunk 25]
```diff
@@ -42,7 +44,9 @@ pub trait Approved<Balance> {
 }
 
 /// Return `true` iff `n1 / d1 < n2 / d2`. `d1` and `d2` may not be zero.
-fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy>(mut n1: T, mut d1: T, mut n2: T, mut d2: T) -> bool {
+fn compare_rationals<T>(mut n1: T, mut d1: T, mut n2: T, mut d2: T) -> bool
+	where T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T, Output = T> + Ord + Copy
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]: suggest this be included in a separate PR since it changes the implementation (instead of mixed into a documentation-specific PR)

File: srml/democracy/src/vote_threshold.rs [hunk 26]
```diff
@@ -69,14 +73,38 @@ fn compare_rationals<T: Zero + Mul<T, Output = T> + Div<T, Output = T> + Rem<T,
 	}
 }
 
-impl<Balance: IntegerSquareRoot + Zero + Ord + Add<Balance, Output = Balance> + Mul<Balance, Output = Balance> + Div<Balance, Output = Balance> + Rem<Balance, Output = Balance> + Copy> Approved<Balance> for VoteThreshold {
-	/// Given `approve` votes for and `against` votes against from a total electorate size of
-	/// `electorate` of whom `voters` voted (`electorate - voters` are abstainers) then returns true if the
-	/// overall outcome is in favor of approval.
+impl<Balance> Approved<Balance> for VoteThreshold
+	where Balance: IntegerSquareRoot
```
Comment (ltfschoen) [inline srml/democracy/src/vote_threshold.rs]: suggest this be included in a separate PR since it changes the implementation (instead of mixed into a documentation-specific PR)

**Comment by ltfschoen** [comment]
With regard to the actual implementation, I came up with the following questions:

1) In the implementation of `set_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R305, how are we checking that it is actually being called by a **stash account** as expected (rather than being called by a **proxy account** for instance)? do we need an additional assertion?

2) In the implementation of `resign_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6R312, how are we checking that it is actually being called by a **proxy account** as expected (rather than being called by a **stash account** for instance)? do we need an additional assertion?

3) In the implementation of `remove_proxy` https://github.com/paritytech/substrate/pull/2286/files#diff-cbe2b253de4fb4f87ccc41b9075568d6L176, how are we checking that it is actually being called by a **stash account** as expected (rather than being called by a **proxy account** for instance)? do we need an additional assertion?

File: srml/democracy/src/lib.rs (hunk 11 above)
Comment (shawntabrizi) [reply srml/democracy/src/lib.rs]: I think there are spacing issues in GitHub which do not appear when you open it in Visual Studio Code.

File: srml/democracy/src/vote_threshold.rs (hunk 25 above)
Comment (shawntabrizi) [reply srml/democracy/src/vote_threshold.rs]: It does not change anything really, it is just a reformat of the code

File: srml/democracy/src/lib.rs [hunk 27]
```diff
@@ -14,7 +14,143 @@
...
+//! 			Ok(())
+//! 		}
+//! 	}
+//! }
+//! # fn main(){}
+//! ```
+//!
+//! ## Genesis Config
+//!
+//! The Democracy module depends on the [`GenesisConfig`](./struct.GenesisConfig.html).
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply srml/democracy/src/lib.rs]: Sounds good

File: srml/democracy/src/lib.rs [hunk 28]
```diff
@@ -14,7 +14,143 @@
...
+//!  the winning proposal. This proposal is removed from `PublicProps` and becomes a public referendum.
+//!  The accounts that locked a deposit into this winning proposal are refunded their reserved deposit.
+//!
+//! #### Referenda Terminology
+//!
+//! - **Start:** Start a public referendum using the `start_referendum` call that anyone may execute
+//!  by signing and submitting an extrinsic. It is allocated the next referendum index, which is mapped to its
+//!  corresponding voting period expiry block, the proposal it relates to, and the given voting threshold.
+//! - **Cancellation:** Remove all information about a referendum.
+//! - **Validity:** The new referendum must not have a voting period that ends before any existing referenda.
```
[the remark this answers is not in the document]
Comment (shawntabrizi) [reply srml/democracy/src/lib.rs]: No, the current line makes sense based on the code.

		if ref_index > 0 && Self::referendum_info(ref_index - 1).map(|i| i.end > end).unwrap_or(false) {
			Err("Cannot inject a referendum that ends earlier than preceeding referendum")?
		}
